### PR TITLE
Achievements list (fetching from PGS) 

### PR DESCRIPTION
**Pull request by Daniel-ND**

### CODE REVIEW

File: app/src/main/java/com/google/droidjump/AchievementsFragment.java [hunk 1]
```diff
@@ -50,57 +51,17 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
...
-            public void onBindViewHolder(AchievementViewHolder viewHolder, int i) {
+        activity.getAchievementsClient().load(true).addOnCompleteListener(activity, task -> {
+            if (task.isSuccessful()) {
+                AchievementBuffer achievementBuffer = task.getResult().get();
+                if (achievementBuffer != null) {
+                    AchievementsAdapter adapter = new AchievementsAdapter(achievementBuffer, activity);
+                    achievementsView.setAdapter(adapter);
+                }
+            } else {
+                Log.e("AchievementsFragment", "Failed to load achievements from client");
```
Comment (redls) [inline app/src/main/java/com/google/droidjump/AchievementsFragment.java:62]: nit: say what happen to trigger the error.

File: app/src/main/res/layout/section.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="48dp"
+    android:gravity="center_vertical"
+    android:paddingLeft="16dp"
+    android:singleLine="true"
```
Comment (redls) [inline app/src/main/res/layout/section.xml:23]: can we maybe have these numbers as dimensions?

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        this.achievementBuffer = achievementBuffer;
+        this.activity = activity;
+        fillViewItems();
+    }
+
+    private void fillViewItems() {
+        items = new ArrayList<>();
+        sectionNames = new ArrayList<>();
+        for (int i = 0; i < achievementBuffer.getCount(); i++) {
+            if (i == 0 && achievementBuffer.get(i).getState() == Achievement.STATE_UNLOCKED) {
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:53]: Could we put if (i == 0) separately to not check it twice?

Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:53]: Extract 
```
achievementBuffer.get(i).getState() == Achievement.STATE_UNLOCKED
```
into separate variable to use it here and below.

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                items.add(new RecyclerViewItem(ItemType.SECTION_NAME, sectionNames.size()));
+                sectionNames.add(String.valueOf(R.string.unlocked_section_name));
+            } else if (i == 0 || !(achievementBuffer.get(i).getState() == Achievement.STATE_UNLOCKED) && achievementBuffer.get(i - 1).getState() == Achievement.STATE_UNLOCKED) {
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:56]: Would also suggest to extract 
```
achievementBuffer.get(i - 1).getState() == Achievement.STATE_UNLOCKED
```
into separate variable to make the if clause readable - currently is it too long.

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                sectionNames.add(String.valueOf(R.string.locked_section_name));
+            }
+            items.add(new RecyclerViewItem(ItemType.ACHIEVEMENT, i));
+        }
+    }
+
+    @NonNull
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        if (viewType == 0) {
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:67]: What is a "viewType == 0"? Is there some enum/constant we could compare with instead?

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        ImageView icon = holder.getIcon();
+        ImageManager manager = ImageManager.create(activity);
+        switch (achievement.getState()) {
+            case Achievement.STATE_UNLOCKED:
+                manager.loadImage(icon, achievement.getUnlockedImageUri());
+                break;
+            case Achievement.STATE_REVEALED:
+                manager.loadImage(icon, achievement.getRevealedImageUri());
+                break;
+            case Achievement.STATE_HIDDEN:
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:95]: As discussed, might be worth to actually load an image of a question mark. Might be in the future PR, not necessarily here.

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                break;
+        }
+    }
+
+    @Override
+    public int getItemViewType(int position) {
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:101]: Can this return an enum or a constant? Currently 0 and 1 does not mean anything.

**Review by saimonsaret — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java (hunk 3 above)
Comment (Daniel-ND) [reply app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:53]: Done

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java (hunk 6 above)
Comment (Daniel-ND) [reply app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:107]: Sure, I don't know where to get such a picture yet, so I'll leave this task for the next pull requests.

File: app/src/main/java/com/google/droidjump/AchievementsFragment.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Daniel-ND) [reply app/src/main/java/com/google/droidjump/AchievementsFragment.java]: Added toast, think for now it’s enough

File: app/src/main/java/com/google/droidjump/AchievementsFragment.java [hunk 8]
```diff
@@ -50,57 +53,19 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
...
+            if (task.isSuccessful()) {
+                AchievementBuffer achievementBuffer = task.getResult().get();
+                if (achievementBuffer != null) {
+                    AchievementsAdapter adapter = new AchievementsAdapter(achievementBuffer, activity);
+                    achievementsView.setAdapter(adapter);
+                }
+            } else {
+                String message = Objects.requireNonNull(task.getException()).getMessage();
+                Log.e("AchievementsFragment", "Failed to load achievements from client: " + message);
+                Toast.makeText(activity, "Oops, something went wrong", Toast.LENGTH_SHORT).show();
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/AchievementsFragment.java:66]: Let's leave a todo to improve this behaviour

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            int currentAchievementState = achievementBuffer.get(i).getState();
+            if (i == 0) {
+                if (currentAchievementState == Achievement.STATE_UNLOCKED) {
+                    items.add(new RecyclerViewItem(ItemType.SECTION_NAME, sectionNames.size()));
+                    sectionNames.add(activity.getString(R.string.unlocked_section_name));
+                } else {
+                    items.add(new RecyclerViewItem(ItemType.SECTION_NAME, sectionNames.size()));
+                    sectionNames.add(activity.getString(R.string.locked_section_name));
+                }
+            } else if (!(currentAchievementState == Achievement.STATE_UNLOCKED)) {
```
Comment (saimonsaret) [inline app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:65]: Extract 
```
currentAchievementState == Achievement.STATE_UNLOCKED
```
to a variable - this value is used twice here and above.

File: app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java (hunk 6 above)
Comment (saimonsaret) [reply app/src/main/java/com/google/droidjump/achievements_data/AchievementsAdapter.java:107]: Let's leave a TODO here then.

**Review by saimonsaret — APPROVED** [review]

**Review by saimonsaret — APPROVED** [review]

**Pull request merged**